---
title: "MSSM-Penn-Pitt - ACC and DLPFC - DREAM - FEATURECOUNTS - GRCh38 - CQN - Sparse Model"
author: "Kelsey Montgomery, Gabriel Hoffman, Thanneer Perumal"
date: "`r date()`"
output: html_document
---

```{r knit2synapse, eval=FALSE}
library(synapser)
library(knit2synapse)

synLogin()

knit2synapse::createAndKnitToFolderEntity(file = "./cmc-rnaseq-de/MSSM-Penn-Pitt_ACC_DLPFC_FEATURECOUNTS_GRCh38_CQN_Sparse.Rmd",
                                 parentId ="syn21435176",
                                 folderName = "MSSM-Penn-Pitt - ACC and DLPFC - DREAM - FEATURECOUNTS - GRCh38 - CQN")
```

```{r libs, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## Load required libraries
library(CovariateAnalysis) # get the package from devtools::install_github('th1vairam/CovariateAnalysis@dev')
library(data.table)
library(tidyr)
library(plyr)
library(dplyr)
library(stringr)
library(R.utils)
library(ggplot2)
library(limma)
library(psych)
library(edgeR)
library(cqn)
library(lme4)
library(biomaRt)
library(variancePartition)
library(synapser)
library(knitr)
library(githubr) # get the package from devtools::install_github('brian-bot/githubr')

synLogin()

library(doParallel)
library(foreach)
library(future)

cl = makeCluster((availableCores()-2)/2)

registerDoParallel(cl)

options(xtable.type="html")

knitr::opts_chunk$set(
  echo=FALSE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  tidy = FALSE,
  cache = TRUE)
```

```{r synapse.parameters, include=FALSE, cache=TRUE}
parentId = 'syn9872607';
activityName = 'Covariate adjustments';
activityDescription = 'Covariate analysis of GRCh38 feature counts with CQN normalisation (ACC and DLPFC)';

thisFileName <- 'MSSM-Penn-Pitt_ACC_DLPFC_FEATURECOUNTS_GRCh38_CQN_Sparse.Rmd'

# Github link
thisRepo <- getRepo(repository = "kelshmo/analysis", ref="branch", refName='add-dream')
thisFile <- getPermlink(repository = thisRepo, repositoryPath=paste0('cmc-rnaseq-de/',thisFileName))
```
### Data download
#### Obtain count matrix and metadata from synapse.
```{r download.data, cache=TRUE}
downloadFile <- function(id){
  fread(synGet(id)$path, data.table = F)
}
downloadFile_version <- function(id , version){
  fread(synGet(id, version = version)$path, data.table = F)
}
# Download reprocessed counts
counts = 'syn21754352'
ALL_USED_IDs = counts
counts = downloadFile(counts) %>% data.frame()
counts$Chr <- NULL
counts$Start <- NULL
counts$End <- NULL
counts$Strand <- NULL

# Parse gene lengths
gene_lengths = counts[,c("Geneid", "Length")]
counts$Length <- NULL
gene_lengths <- dplyr::rename(gene_lengths, gene_id = Geneid)

# Get ancestry vector calculated using gemtools
ANCESTRY_ID = 'syn2511399'
ALL_USED_IDs[length(ALL_USED_IDs)+1] = ANCESTRY_ID
ANCESTRY = downloadFile(ANCESTRY_ID) %>% 
  plyr::rename(c('DNA_report..Genotyping.Sample_ID' = 'SNP_report:Genotyping_Sample_ID'))

# Get genotype ids from synapse
GENOTYPE_ID = 'syn16816490'
ALL_USED_IDs = c(ALL_USED_IDs, GENOTYPE_ID)
GENOTYPE = downloadFile(GENOTYPE_ID) %>% 
  dplyr::select(Individual_ID, `SNP_report:Genotyping_Sample_ID`) %>%
  dplyr::inner_join(ANCESTRY)

#Get Metadata
metadata = 'syn16816488' 
ALL_USED_IDs[length(ALL_USED_IDs)+1] = metadata
metadata = downloadFile_version(metadata, version = 14)  %>% 
  dplyr::select(Individual_ID, Institution, Reported_Gender, Sex, 
                Ethnicity, ageOfDeath, `PMI_(in_hours)`, Dx,
                Sample_RNA_ID, one_of('rnaSeq_isolation:RIN',
                                      'rnaSeq_report:Ribozero_Batch', 'rnaSeq_report:Library_Batch', 
                                      'rnaSeq_report:Flowcell_Batch','rnaSeq_dissection:Brain_Region',
                                      'rnaSeq_report:Exclude?',
                                      'rnaSeq_report:Mapped_Reads',
                                      'rnaSeq_report:Intragenic_Rate','rnaSeq_report:Intronic_Rate',
                                      'rnaSeq_report:Intergenic_Rate','rnaSeq_report:Genes_Detected',
                                      'rnaSeq_report:Expression_Profiling_Efficiency',
                                      'rnaSeq_report:rRNA_Rate',
                                      'rnaSeq_report:Total_Reads', 'rnaSeq_report:Percent_Aligned'))

METADATA = metadata %>%
  dplyr::left_join(GENOTYPE) %>%
  dplyr::rename(PMI = `PMI_(in_hours)`,
                ReportExclude = `rnaSeq_report:Exclude?`,
                SampleID = Sample_RNA_ID,
                LibraryBatch = `rnaSeq_report:Library_Batch`,
                FlowcellBatch = `rnaSeq_report:Flowcell_Batch`,
                RibozeroBatch = `rnaSeq_report:Ribozero_Batch`,
                Tissue = `rnaSeq_dissection:Brain_Region`,
                MappedReads = `rnaSeq_report:Mapped_Reads`,
                IntragenicRate = `rnaSeq_report:Intragenic_Rate`, 
                IntronicRate = `rnaSeq_report:Intronic_Rate`, 
                IntergenicRate = `rnaSeq_report:Intergenic_Rate`, 
                GenesDetected = `rnaSeq_report:Genes_Detected`,
                ExpProfEfficiency = `rnaSeq_report:Expression_Profiling_Efficiency`, 
                rRNARate = `rnaSeq_report:rRNA_Rate`,
                TotalReads = `rnaSeq_report:Total_Reads`, 
                AlignmentRate = `rnaSeq_report:Percent_Aligned`,
                RIN = `rnaSeq_isolation:RIN`) %>%
  dplyr::select(SampleID, Individual_ID, Institution, Reported_Gender,Sex, Ethnicity, ageOfDeath, PMI, Dx, 
                RIN, ReportExclude, EV.1, EV.2, EV.3, EV.4, EV.5,
                LibraryBatch, FlowcellBatch, RibozeroBatch, Tissue, MappedReads, TotalReads, GenesDetected, AlignmentRate,
                IntragenicRate, IntergenicRate, IntronicRate, ExpProfEfficiency, rRNARate) %>% 
  dplyr::filter(Institution %in% c("MSSM", "Penn", "Pitt")) %>% 
  dplyr::filter(SampleID %in% colnames(counts), !is.na(SampleID)) %>%
  dplyr::mutate(Dx = forcats::fct_recode(Dx,  Other = "AFF", Other = "BP", Control = "Control", SCZ = "SCZ"))
```

### Data preprocessing
```{r preprocess.data}
ind = METADATA$SampleID[which(METADATA$ReportExclude == 1)]
writeLines(paste('Following',length(ind),'samples are marked exclude'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Ethnicity) | is.na(METADATA$Institution) | is.na(METADATA$Dx)]
writeLines(paste('Following', length(ind), 'counts are missing clinical metadata'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$PMI)]
writeLines(paste('Following', length(ind), 'counts are missing PMI'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$Reported_Gender)]
writeLines(paste('Following', length(ind), 'counts are missing gender'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 

ind = METADATA$SampleID [is.na(METADATA$ageOfDeath)]
writeLines(paste('Following', length(ind), 'counts are missing age of death'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))

ind = METADATA$SampleID [is.na(METADATA$EV.1)]
writeLines(paste('Following', length(ind), 'counts are missing ancestry information'))
writeLines(paste(ind, collapse = ', '))
METADATA <- METADATA  %>% dplyr::filter(!(SampleID %in% ind))
```

```{r preprocess.data1, results='asis'}
# Match covariates to expression data
indToRetain = intersect(METADATA$SampleID, colnames(counts))

rownames(counts) <- counts$Geneid
counts = counts[,indToRetain]

rownames(METADATA) = METADATA$SampleID
METADATA = METADATA[indToRetain,]

METADATA %>% 
  group_by(Dx, Tissue) %>% 
  dplyr::summarise(count = n()) %>% 
  spread(Tissue, count) %>%
  kable()

METADATA = METADATA %>%
  dplyr::mutate(Dx.Tissue = paste(Dx, Tissue, sep = '.'))
```

```{r gene.param}
## Get GC content from biomart
backgroundGenes = data.frame(gene_id = rownames(counts)) %>%
  dplyr::mutate(id = gene_id) %>%
  tidyr::separate(id, c('ensembl_gene_id','position'), sep = '\\.')

# Define biomart object
mart <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                host = "uswest.ensembl.org", # Ensembl Release 99 (January 2020)
                dataset = "hsapiens_gene_ensembl")

# Query biomart
Ensemble2HGNC <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "percentage_gene_gc_content", "gene_biotype", "chromosome_name"),
                       filters = "ensembl_gene_id", values = backgroundGenes$ensembl_gene_id,
                       mart = mart)

GENE.GC.CONT = Ensemble2HGNC %>%
  dplyr::left_join(backgroundGenes) %>% 
  dplyr::select(gene_id, percentage_gene_gc_content, chromosome_name) %>%
  unique
rownames(GENE.GC.CONT) = GENE.GC.CONT$gene_id
```

### Sex Check
```{r Reported_Gender_plots}
counts$gene_id = rownames(counts) 
REPORTED.GENDER.COUNTS = GENE.GC.CONT %>% 
  left_join(counts) %>%
  dplyr::select(-one_of("percentage_gene_gc_content")) %>%
  filter(chromosome_name == "X" |chromosome_name == "Y") %>% 
  tidyr::gather(key = item, value = value, -c(gene_id, chromosome_name)) %>%
  dplyr::mutate(value = log(value)) %>%
  dplyr::rename(`counts(log)`= value) %>% 
  dplyr::rename(SampleID = item) %>%
  left_join(METADATA[,c("SampleID", "Reported_Gender")]) %>% 
  dplyr::rename(`Reported Gender` = Reported_Gender) 

##XIST and UTY expression 
#ENSG00000229807.11 and ENSG00000183878.15 
FILT <- REPORTED.GENDER.COUNTS %>% 
  filter(gene_id == "ENSG00000229807.11" | gene_id == "ENSG00000183878.15") %>% 
  dplyr::select(-one_of("chromosome_name")) %>% 
  tidyr::spread(key = gene_id, value = `counts(log)`) %>% 
  mutate(XIST = as.numeric(`ENSG00000229807.11`)) %>% 
  mutate(UTY = as.numeric(`ENSG00000183878.15`))

p = ggplot(FILT, aes (x= XIST, y = UTY)) 
p = p + geom_point(aes(color=`Reported Gender`)) 
p
```
### Covariates clustering
Determine relationship between covariates. 
```{r covariates.clustering}
FactorCovariates <- c('Individual_ID', "Institution", "Reported_Gender", "LibraryBatch", "Dx.Tissue", "RibozeroBatch", "FlowcellBatch")
ContCovariates <- c("ageOfDeath", "PMI", "RIN", "MappedReads", "IntragenicRate", "IntronicRate", "IntergenicRate",
                    "GenesDetected", "ExpProfEfficiency", "rRNARate", "TotalReads","AlignmentRate", 
                     "EV.1", "EV.2", "EV.3", "EV.4", "EV.5")

# Find inter relation between factor covariates
COVARIATES = METADATA[,c(FactorCovariates,ContCovariates),drop=F]
COVARIATES[,FactorCovariates] <- data.frame(lapply(COVARIATES[,FactorCovariates],function(x){
  x <- sapply(x,function(y){str_replace_all(as.character(y),'[^[:alnum:]]','_')})}))
rownames(COVARIATES) <- METADATA$SampleID

# Convert factor covariates to factors
COVARIATES[,FactorCovariates] = lapply(COVARIATES[,FactorCovariates], factor)
COVARIATES[,ContCovariates] = lapply(COVARIATES[,ContCovariates], function(x){
  x = as.numeric(as.character(gsub('[\\,\\%]','',x)))
})

# Add in RIN^2 values
COVARIATES$RIN2 = COVARIATES$RIN^2
ContCovariates = c(ContCovariates, 'RIN2')
```
Correlation/association between covariates at an FDR <= 0.1
```{r covariates.correlation, fig.width=10, fig.height=8}
COVARIATES.CORRELATION = getAssociationStatistics(COVARIATES, PVAL = 0.05)
tmp = COVARIATES.CORRELATION$ESTIMATE
tmp[COVARIATES.CORRELATION$PVAL > 0.05] = 0
h = Heatmap(tmp, col = colorRamp2(c(-1,0,1), c('blue','white','red')), name = 'AssocEstimate')
ComplexHeatmap::draw(h, heatmap_legend_side = 'left')
```

## Explore metatdata
```{r data.explore, fig.width = 12, fig.height = 12}
my.theme <- theme_bw() %+replace% theme(legend.position = 'top', axis.text.x = element_text(angle = 90, hjust = 1), plot.title=element_text(hjust=0.5))

# RIN
p = list()
p[[1]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = RIN)) + geom_boxplot()
p[[1]] = p[[1]] + ggtitle('RIN') + my.theme

# Age of Death
p[[2]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = ageOfDeath)) + geom_boxplot()
p[[2]] = p[[2]] + ggtitle('AgeOfDeath') + my.theme

# PMI
p[[3]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = PMI)) + geom_boxplot()
p[[3]] = p[[3]] + ggtitle('PMI (in hours)') + my.theme

# Intronic Rate
p[[4]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntronicRate)) + geom_boxplot()
p[[4]] = p[[4]] + ggtitle('Intronic Rate') + my.theme

# IntergenicRate
p[[5]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntergenicRate)) + geom_boxplot()
p[[5]] = p[[5]] + ggtitle('Intergenic Rate') + my.theme

# Transcripts Detected
p[[6]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = IntragenicRate)) + geom_boxplot()
p[[6]] = p[[6]] + ggtitle('Intragenic Rate') + my.theme

# Mapped Reads
p[[7]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = MappedReads)) + geom_boxplot()
p[[7]] = p[[7]] + ggtitle('Mapped Reads') + my.theme

# PercentAligned
p[[8]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = TotalReads)) + geom_boxplot()
p[[8]] = p[[8]] + ggtitle('Total Reads') + my.theme

# rRNARate
p[[9]] = ggplot(COVARIATES, aes(x = Dx.Tissue, y = rRNARate)) + geom_boxplot()
p[[9]] = p[[9]] + ggtitle('rRNARate') + my.theme

multiplot(plotlist = p, cols = 3)

# Institution
# vcd::mosaic(~ Institution + Dx.Tissue, data = COVARIATES)

# Gender
# vcd::mosaic(~ Gender + Dx.Tissue, data = COVARIATES)
```
### Filter genes
Remove genes that have less than 1 cpm counts in at least 50% of samples per Dx.Tissue. Also remove genes with missing gene length and percentage GC content
```{r cpmnormalisation}
counts$gene_id = NULL

genesToAnalyze = dlply(METADATA, .(Dx.Tissue, Reported_Gender), .fun = function(mtd, count){
  processed.counts = getGeneFilteredGeneExprMatrix(count[,mtd$SampleID],
                                                   MIN_GENE_CPM=1, 
                                                   MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0.5)
  processed.counts$filteredExprMatrix$genes
}, counts)

genesToAnalyze = unlist(genesToAnalyze) %>% 
  unique() %>%
  intersect(GENE.GC.CONT$gene_id[!is.na(GENE.GC.CONT$percentage_gene_gc_content)]) %>%
  intersect(gene_lengths$gene_id[!is.na(gene_lengths$Length)])

#Subset union of all genes and convert subset to logcpm values 
PROCESSED_COUNTS = getGeneFilteredGeneExprMatrix(counts[genesToAnalyze, ], MIN_GENE_CPM=0, MIN_SAMPLE_PERCENT_WITH_MIN_GENE_CPM=0)
pct.pc = PROCESSED_COUNTS$filteredExprMatrix$genes %>%
  dplyr::rename(gene_id = genes) %>%
  left_join(backgroundGenes) %>%
  left_join(Ensemble2HGNC) %>%
  group_by(gene_biotype) %>%
  dplyr::summarise(fraction  = n()) %>%
  filter(fraction > 100) %>%
  dplyr::mutate(fraction = fraction/length(PROCESSED_COUNTS$filteredExprMatrix$genes[,1]))
```
`r dim(PROCESSED_COUNTS$filteredExprMatrix$genes)[1]` genes are used for the analysis. 

Following is the distribution of biotypes
`r kable(pct.pc)`

### Outlier analysis
Detect outlier samples based on their expression (logCPM) pattern
```{r detect.outliers, fig.height=8, fig.width=8, results='asis', cache = FALSE}
indToRemove = c('MSSM_RNA_ACC_BP_24', 'MSSM_RNA_ACC_375', 'MSSM_RNA_ACC_170', 'MSSM_RNA_ACC_315', 'MSSM_RNA_ACC_250',
                'MSSM_RNA_ACC_21', 'MSSM_RNA_ACC_356', 'MSSM_RNA_ACC_318', 'MSSM_RNA_ACC_222', 'MSSM_RNA_ACC_220',
                'MSSM_RNA_ACC_379', 'MSSM_RNA_PFC_133', 'MSSM_RNA_PFC_139', 'MSSM_RNA_PFC_163', 'MSSM_RNA_PFC_260',
                'MSSM_RNA_PFC_299', 'MSSM_RNA_PFC_361', 'MSSM_RNA_PFC_89', 'MSSM_RNA_PFC_319', 'MSSM_RNA_PFC_348',
                'MSSM_RNA_PFC_354', 'MSSM_RNA_PFC_60', 'PENN_RNA_PFC_71', 'PENN_RNA_PFC_56',  "MSSM_RNA_ACC_310",
                "MSSM_RNA_ACC_51",  "MSSM_RNA_ACC_254", "MSSM_RNA_ACC_28", "MSSM_RNA_ACC_293", "MSSM_RNA_ACC_233",
                "MSSM_RNA_ACC_167", "MSSM_RNA_ACC_353", "MSSM_RNA_ACC_355", "MSSM_RNA_ACC_211", "PENN_RNA_ACC_122",
                "PENN_RNA_ACC_049", "PENN_RNA_ACC_052", "PENN_RNA_ACC_068", "PENN_RNA_ACC_070", "PITT_RNA_ACC_10020",
                "PITT_RNA_ACC_1240", "PITT_RNA_ACC_1284", "PITT_RNA_ACC_700", "PITT_RNA_ACC_857", "PITT_RNA_ACC_BP_1589",
                "MSSM_RNA_PFC_230", "MSSM_RNA_PFC_304", "MSSM_RNA_PFC_309", "MSSM_RNA_PFC_329", "MSSM_RNA_PFC_82",
                "PENN_RNA_PFC_5", "PENN_RNA_PFC_58", "PENN_RNA_PFC_99", "PITT_RNA_PFC_1284")

indToRemove = c(indToRemove, c("MSSM_RNA_ACC_224", "MSSM_RNA_ACC_240", "MSSM_RNA_ACC_BP_11", "MSSM_RNA_ACC_42", "MSSM_RNA_ACC_BP_15", "PENN_RNA_ACC_003", "MSSM_RNA_ACC_374"))

indToRemove = c(indToRemove, "PITT_RNA_BP_PFC_1770")

# Find principal components of expression to plot
PC <- prcomp(voom(PROCESSED_COUNTS$filteredExprMatrix$counts)$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

# Percentage from each PC
eigen <- PC$sdev^2
pc1 <- eigen[1]/sum(eigen)
pc2 <- eigen[2]/sum(eigen)

# Identify outliers - samples 4SDs from the mean
outliers <- as.character(plotdata$SampleID[c(which(plotdata$PC1 < mean(plotdata$PC1) - 4*sd(plotdata$PC1)),
                              which(plotdata$PC1 > mean(plotdata$PC1) + 4*sd(plotdata$PC1))), drop = T])

outliers <- c(outliers, as.character(plotdata$SampleID[c(which(plotdata$PC2 < mean(plotdata$PC2) - 4*sd(plotdata$PC2)),
                                           which(plotdata$PC2 > mean(plotdata$PC2) + 4*sd(plotdata$PC2))), drop = T] ))
  
plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, "SampleID")) %>%
    tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_') %>%
  dplyr::mutate(label = SampleID) %>% 
  dplyr::mutate(label = ifelse((label %in% outliers), label, NA))

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=ageOfDeath))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~.)
p <- p + geom_text(aes(label= label), size=4, hjust=0)
p

# Plot abberent distribution of logcpm counts
tmp1 = voom(PROCESSED_COUNTS$filteredExprMatrix$counts)$E %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  tidyr::gather(SampleID, logCPM, -ensembl_gene_id) %>%
  left_join(COVARIATES %>%
              rownameToFirstColumn('SampleID') %>%
              tidyr::separate(Dx.Tissue, c('Dx', 'Tissue'), sep = '_'))

p = ggplot(tmp1 %>%
             dplyr::filter(SampleID %in% outliers),
           aes(x = logCPM, color = SampleID)) + geom_density() 
p = p + theme(legend.position = 'top') + facet_grid(Dx+.~Tissue, scale = 'free')
p

indToRetain = setdiff(colnames(PROCESSED_COUNTS$filteredExprMatrix$counts), outliers)
PROCESSED_COUNTS$filteredExprMatrix$counts = PROCESSED_COUNTS$filteredExprMatrix$counts[,indToRetain]
COVARIATES = COVARIATES[indToRetain,]

tmp = tidyr::separate(COVARIATES, Dx.Tissue, c('Dx', 'Tissue'), sep = '_') %>%
  group_by(Dx, Tissue) %>%
  dplyr::summarise(count = n()) %>%
  spread(Tissue, count)

NEW.COUNTS = PROCESSED_COUNTS$filteredExprMatrix$counts
```
Processing `r dim(PROCESSED_COUNTS$filteredExprMatrix)[1]` genes in `r dim(PROCESSED_COUNTS$filteredExprMatrix)[2]` samples from `r length(unique(COVARIATES$Individual_ID))` unique individuals

Based on the expression pattern following samples were tagged as outliers: `r paste(outliers, collapse = ', ')` 

Based on prior evidence, investigators elected to remove: `r setdiff(indToRemove, outliers)[setdiff(indToRemove,outliers) %in% colnames(counts)]`

Distribution of samples are:
`r kable(tmp)`

### Library Normalisation
Initial library normalisation usign cqn
```{r cqn}
# Compute offset for gene length and gc content
gene_lengths = tibble::column_to_rownames(gene_lengths, var = 'gene_id') %>% as.data.frame()
CQN.GENE_EXPRESSION = cqn(NEW.COUNTS, 
                          x = GENE.GC.CONT[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'percentage_gene_gc_content'],
                          lengths = gene_lengths[PROCESSED_COUNTS$filteredExprMatrix$genes$genes, 'Length'],
                          lengthMethod = "smooth", 
                          verbose = FALSE)
CQN.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$y + CQN.GENE_EXPRESSION$offset
```

### Co-expression distribution
Coexpression of genes 
```{r coexp1, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(CQN.GENE_EXPRESSION$E))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Sample clustering
```{r decompse.normalise.data, fig.height=8, fig.width=8, results='asis', cache=FALSE}
# Find principal components of expression to plot
PC <- prcomp(CQN.GENE_EXPRESSION$E, scale.=T, center = T)

# Plot first 2 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_')

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=ageOfDeath))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~.)
# p <- p + geom_text(aes(label= SampleID), size=4, hjust=0)
p
```
Tree based clustering of samples
```{r decompse.normalise.data.1, fig.height=6, fig.width=10, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c("Institution", "Reported_Gender", "LibraryBatch", "Dx.Tissue")])
COVARIATES.tmp[is.na(COVARIATES.tmp)] = 0

tree = hclust(as.dist(t(CQN.GENE_EXPRESSION$E)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp))
```
```{r temp, include = F}
dev.off()
gc()
```

### Distribution of samples (log cpm)
```{r lcpm.dist, cache=FALSE}
# Plot abberent distribution of logcpm counts
tmp1 = (CQN.GENE_EXPRESSION$E) %>%
  rownameToFirstColumn('gene_id') %>%
  tidyr::gather(SampleID, logCPM, -gene_id) %>%
  left_join(COVARIATES %>%
              rownameToFirstColumn('SampleID'))

p = ggplot(tmp1, aes(x = logCPM, color = SampleID)) + geom_density() 
p = p + theme_bw() %+replace% theme(legend.position = 'NONE') + facet_grid(.~Dx.Tissue, scale = 'free')
p
```

### Significant Covariates
Correlation between pca of unadjusted mRNA expression and covariates is used to find significant covariates
```{r preAdjusted.covariates, cache=TRUE}
# Find correlation between PC's of gene expression with covariates
preAdjustedSigCovars = runPCAandPlotCorrelations(CQN.GENE_EXPRESSION$E, 
                                                 COVARIATES,
                                                 'NULL design(voom-normalized)', 
                                                 isKeyPlot=TRUE, 
                                                 MIN_PVE_PCT_PC = 1)
```

Significant covariates to adjust at FDR 0.1 are `r preAdjustedSigCovars$significantCovars`
```{r preAdjustedSigCovars.NULL, fig.width=20, fig.height=12}
preAdjustedSigCovars[["PC_res"]][[2]]$plotData
```

#### Model Selection
```{r set_vars}
adjust.covars <- c("Dx.Tissue", "RIN", "IntronicRate")
```

[Code](https://github.com/kelshmo/analysis/blob/f7345adde2052afbc258773f69930df300989bf3/cmc-rnaseq-de/MSSM-Penn-Pitt_ACC_DLPFC_FEATURECOUNTS_GRCh38_CQN_Sparse.Rmd#L579-L737) for model selection and [BIC Criteria](https://www.synapse.org/#!Synapse:syn21543160)

#### Variables included in the model 
Covariates were added as fixed effects iteratively in three phases if model improvement by BIC criteria was observed in the majority of genes. Clinical, ancestry and sample-specific technical variables were tested for model improvement first. Covariates related to batch effects were next tested for model improvement and finally, RNASeq alignment-specific covariates.

Clinical, technical and batch variables selected after phase I, II and III of model identification are `r paste(adjust.covars, collapse = ', ')`

### Normalisation - Dx.Tissue
1. Dx.Tissue is chosen as the primary variable of interest (i.e., covariate adjustments is conditioned on diagnosis and tissue)
2. Individual_ID is chosen as random effect
```{r iterative.normalisation, results='asis'}
writeLines(paste('Using following covariates in the model:',
                 paste(adjust.covars, collapse=', '),
                 'as fixed effects and Individual_ID is chosen as random effect'))

# Post adjusted formula
random_effect <- "Individual_ID"
formula <- glue::glue("~ ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")

# Estimate voom weights with DREAM
geneExpr = DGEList(NEW.COUNTS)
geneExpr = edgeR::calcNormFactors(geneExpr)

VOOM.GENE_EXPRESSION = variancePartition::voomWithDreamWeights(counts = geneExpr, 
                                                               formula = formula,
                                                               data = COVARIATES,
                                                               save.plot = TRUE)

# How to plot a voom object if save.plot=TRUE
# Set plot generic to voom result EList
#' @export
setMethod("plot", signature(x="EList", y="missing"),
  function(x,  y, ...){
    if( is.null(x$voom.xy) || is.null(x$voom.line)){
        stop("x does not contain the trend information.\nvoom() must be run with save.plot=TRUE")
    }
    # points
    df = data.frame(x = x$voom.xy$x,
                    y = x$voom.xy$y)
    # trend line
    df.line = data.frame(x = x$voom.line$x,
                        y = x$voom.line$y)
    ggplot(df, aes(x,y)) + geom_point(size=.1) + theme_bw(15) + 
      theme(aspect.ratio=1, plot.title = element_text(hjust = 0.5)) + 
      geom_line(data=df.line, aes(x,y), color="red") + xlab(bquote(log[2](count + 0.5))) + 
      ylab(expression( sqrt("standard deviation"))) + ggtitle("Voom: Mean-variance trend") + 
      ylim(0, max(df$y))
  }
)
  
plot(VOOM.GENE_EXPRESSION)

# Fit linear model using new weights and new design
VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E
Dx.Tissue_ADJUSTED.FIT = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION,
                                        formula = formula,
                                        data = COVARIATES,
                                        computeResiduals = TRUE)

col_type <- dplyr::select(COVARIATES, c(adjust.covars, "Individual_ID")) %>%
  dplyr::summarise_all(class) %>% 
  tidyr::gather(variable, class)

# Model categorical variables as random to report variance fractions
formula_for_variance_fractions = sapply(1:length(col_type$class), function(i){
  switch(col_type$class[i], 
    "factor" = paste0('(1|', col_type$variable[i], ')'),
    "numeric" = col_type$variable[i])
  })

formula_for_variance_fractions <- glue::glue("~ ", glue::glue_collapse(formula_for_variance_fractions, sep = " + "))

violin_plot <- variancePartition::fitExtractVarPartModel(VOOM.GENE_EXPRESSION,
                                                         formula = formula_for_variance_fractions,
                                                         data = COVARIATES)

plotVarPart(sortCols(violin_plot))

# Residuals after normalisation
Dx.Tissue_RESIDUAL.GENE_EXPRESSION = residuals(Dx.Tissue_ADJUSTED.FIT, CQN.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(Dx.Tissue_RESIDUAL.GENE_EXPRESSION, COVARIATES,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```

### Sanity check
```{r residualSigCovars.manual, fig.width=12, fig.height=8}
residualSigCovars[["PC_res"]][[2]]$plotData
```

### Tissue Correlation
```{r tissue_correlation}
hasReps = rownameToFirstColumn(COVARIATES, "SampleID") %>%
  dplyr::select(Individual_ID) %>% 
  table

df_dlpfc = rownameToFirstColumn(COVARIATES, "SampleID") %>% 
  filter( Individual_ID %in% names(hasReps[hasReps==2])) %>%
  filter(grepl("DLPFC", Dx.Tissue)) %>% 
  dplyr::select(SampleID, Individual_ID, Dx.Tissue)
df_acc = rownameToFirstColumn(COVARIATES, "SampleID") %>%
  filter( Individual_ID %in% names(hasReps[hasReps==2])) %>%
  filter(grepl("ACC", Dx.Tissue)) %>% 
  dplyr::select(SampleID, Individual_ID, Dx.Tissue)

df_merge = merge(df_dlpfc, df_acc, by="Individual_ID")

res_std <- rownameToFirstColumn(Dx.Tissue_RESIDUAL.GENE_EXPRESSION, "ensembl_gene_id")

corRes = sapply(res_std$ensembl_gene_id, function(ensGene){

  x = VOOM.GENE_EXPRESSION$E[ensGene,df_merge$SampleID.x,drop=FALSE]
  y = VOOM.GENE_EXPRESSION$E[ensGene,df_merge$SampleID.y,drop=FALSE]

  cor(t(x)[,1], t(y)[,1], use="pairwise.complete.obs")
})

hist(corRes, xlab="Correlation", main="Correlation between ACC and DLPFC sample from the sample individual for each gene from ")

pca = prcomp(t(VOOM.GENE_EXPRESSION$E))

df_pca = merge(pca$x, as.data.frame(rownameToFirstColumn(COVARIATES, "SampleID")), by.x="row.names", by.y="SampleID" ) %>% 
  tidyr::separate(Dx.Tissue, into = c("Dx", "Tissue"), sep = "_")

ggplot(df_pca, aes(PC1, PC2, color=Tissue, shape=Institution)) + geom_point() + theme_bw(17) + theme(aspect.ratio=1)
```

### Residual calculation
Calculate weighted residuals and add back "Dx.Tissue" to the residuals
```{r varsToAddBack}
# Add variable of interest back to the residuals
varsToAddIn = grep("Dx.Tissue", colnames(Dx.Tissue_ADJUSTED.FIT$design), value = T)
Dx.Tissue_RESIDUAL.GENE_EXPRESSION = Dx.Tissue_RESIDUAL.GENE_EXPRESSION +
  Dx.Tissue_ADJUSTED.FIT$coefficients[,varsToAddIn] %*% t(Dx.Tissue_ADJUSTED.FIT$design[,varsToAddIn])
```

### Coexpression of residual expression 
```{r coexp2, cache=FALSE, fig.height=5, fig.width=5}
cr = cor(t(Dx.Tissue_RESIDUAL.GENE_EXPRESSION))
hist(cr, main = 'Distribution of correlation between genes', xlab = 'Correlation')
```

### Clustering residual data
```{r decompse.normalise.data2, fig.height=8, fig.width=8, results='asis'}
# Find principal components of expression to plot
PC <- prcomp(tmp, scale.=T, center = T)

# Plot first 4 PCs
plotdata <- data.frame(SampleID=rownames(PC$rotation), 
                       PC1=PC$rotation[,1], 
                       PC2=PC$rotation[,2])

plotdata <- left_join(plotdata, rownameToFirstColumn(COVARIATES, 'SampleID')) %>%
  tidyr::separate(Dx.Tissue, c('Dx','Tissue'), sep = '_')

p <- ggplot(plotdata, aes(x=PC1, y=PC2))
p <- p + geom_point(aes(color=Institution, shape=Tissue, size=ageOfDeath))
p <- p + theme_bw() + theme(legend.position="right") + facet_grid(Tissue~., scales = 'free_y')
p
```

```{r decompse.normalise.data2.1, fig.height=8, fig.width=12, results='asis'}
# Eucledian tree based analysis
COVARIATES.tmp = data.matrix(COVARIATES[,c('Dx.Tissue','Institution')])

tree = hclust(as.dist(t(tmp)))
cols = WGCNA::labels2colors(COVARIATES.tmp);

WGCNA::plotDendroAndColors(tree, 
                           colors = cols, 
                           dendroLabels = FALSE, 
                           abHeight = 0.80, 
                           main = "Sample dendrogram",
                           groupLabels = colnames(COVARIATES.tmp[,c('Dx.Tissue','Institution')]))
```

```{r temp1, include=F}
dev.off()
gc()
```

### Differential expression analysis (SCZ - Control - Other)
Genes that are differentially expressed at an FDR <= 0.05 are:
```{r set-formula}
formula <- glue::glue("~ 0 + ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")
```

`r formula`

```{r diffExp, fig.height=10, fig.width=15}
L1 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueOther_ACC", "Dx.TissueControl_ACC"))
L2 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueSCZ_ACC", "Dx.TissueControl_ACC"))
L3 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueSCZ_ACC", "Dx.TissueOther_ACC"))
L4 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueOther_DLPFC", "Dx.TissueControl_DLPFC"))
L5 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueSCZ_DLPFC", "Dx.TissueControl_DLPFC"))
L6 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.TissueSCZ_DLPFC", "Dx.TissueOther_DLPFC"))


L = cbind(L1, L2, L3, L4, L5, L6)

contrast <- c(glue::glue_collapse(names(L1[L1 != 0]), "-"), 
              glue::glue_collapse(names(L2[L2 != 0]), "-"),
              glue::glue_collapse(names(L3[L3 != 0]), "-"),
              glue::glue_collapse(names(L4[L4 != 0]), "-"),
              glue::glue_collapse(names(L5[L5 != 0]), "-"),
              glue::glue_collapse(names(L6[L6 != 0]), "-")
)

# Visualize contrast matrix
plotContrasts(L) 

# Fit contrast
FIT.CONTR = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION,
                                     formula = formula,
                                     data = COVARIATES,
                                     L = L)

# Get differential expression
DE = lapply(glue::glue("L{1:6}"), function(i, FIT){
  topTable(FIT, coef=i, number = Inf) %>%
    rownameToFirstColumn('gene_id') %>% 
    left_join(backgroundGenes)
}, FIT.CONTR) 
names(DE) <- contrast

DE = DE %>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Dx.Tissue','',Comparison),
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>%
  tidyr::separate(Comparison, into = c('ref.state','to.state'), sep = '-') %>%
  tidyr::separate(ref.state, into = c('Dx1','Tissue'), sep = '_') %>%
  tidyr::separate(to.state, into = c('Dx2','Tissue'), sep = '_') %>%
  tidyr::unite(Comparison, Dx1, Dx2, sep = '_vs_') %>%
  left_join(Ensemble2HGNC) %>%
  left_join(gene_lengths %>% rownameToFirstColumn('gene_id'))
DE$Direction[DE$adj.P.Val > 0.05 | abs(DE$logFC) < log2(1.2)] = 'NONE'

tmp = DE %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison, Tissue) %>%
  group_by(Comparison, Tissue) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id)))

tmp1 = DE %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, Comparison, Tissue) %>%
  group_by(Comparison, Tissue) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id)))

kable(full_join(tmp,tmp1))

p = ggplot(DE, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme_bw() %+replace% theme(legend.position = 'top')
p = p + facet_grid(Tissue+.~Comparison, scales = 'fixed')
p
```

### Associate differential expression results with gc content, gene length and average expression
```{r associate.de, fig.height=10, fig.width=15}
pl = list()
pl[[1]] = ggplot(DE %>% 
                   arrange(match(Direction, c("NONE", "UP", "DOWN"))), aes(x = log10(Length), y = logFC, color = Direction)) + geom_point() 
pl[[1]] = pl[[1]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = log10(Length), y = logFC))
pl[[1]] = pl[[1]] + facet_grid(ref.Tissue~.) + scale_color_manual(values = c('green','grey','red'))
pl[[1]] = pl[[1]] + theme(legend.position = 'top')

pl[[2]] = ggplot(DE %>% 
                   arrange(match(Direction, c("NONE", "UP", "DOWN"))), aes(x = percentage_gene_gc_content, y = logFC, color = Direction)) + geom_point()
pl[[2]] = pl[[2]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = percentage_gene_gc_content, y = logFC))
pl[[2]] = pl[[2]] + facet_grid(ref.Tissue~.) + scale_color_manual(values = c('green','grey','red'))
pl[[2]] = pl[[2]] + theme(legend.position = 'top')

pl[[3]] = ggplot(DE %>% 
                    arrange(match(Direction, c("NONE", "UP", "DOWN"))), aes(x = AveExpr, y = logFC, color = Direction)) + geom_point()
pl[[3]] = pl[[3]] + geom_smooth(method = 'loess', inherit.aes = FALSE, aes(x = AveExpr, y = logFC))
pl[[3]] = pl[[3]] + facet_grid(ref.Tissue~.) + scale_color_manual(values = c('green','grey','red'))
pl[[3]] = pl[[3]] + theme(legend.position = 'top')

multiplot(plotlist = pl, cols = 3)
```

### Normalisation - Dx.Tissue.Gender
1. Dx.Tissue.Gender is chosen as the primary variable of interest (i.e., covariate adjustments is conditioned on diagnosis, tissue and reported gender)
2. Individual_ID is chosen as random effect
```{r iterative.normalisation-gender, results='asis'}
COVARIATES$Dx.Tissue.Gender = factor(paste(COVARIATES$Dx.Tissue, COVARIATES$Reported_Gender, sep = "_"))
COVARIATES$Reported_Gender <- NULL
COVARIATES$Dx.Tissue <- NULL
adjust.covars[1] <- "Dx.Tissue.Gender"

writeLines(paste('Using following covariates in the model:',
                 paste(adjust.covars, collapse=', '),
                 'as fixed effects and Individual_ID is chosen as random effect'))

# Post adjusted formula
random_effect <- "Individual_ID"
formula <- glue::glue("~ ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")

# Estimate voom weights with DREAM
geneExpr = DGEList(NEW.COUNTS)
geneExpr = edgeR::calcNormFactors(geneExpr)

VOOM.GENE_EXPRESSION = variancePartition::voomWithDreamWeights(counts = geneExpr, 
                                                               formula = formula,
                                                               data = COVARIATES,
                                                               save.plot = TRUE)

plot(VOOM.GENE_EXPRESSION)

# Fit linear model using new weights and new design
VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E
Dx.Tissue.Gender_ADJUSTED.FIT = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION, 
                                        formula = formula,
                                        data = COVARIATES,
                                        computeResiduals = TRUE)

col_type <- dplyr::select(COVARIATES, c(adjust.covars, "Individual_ID")) %>%
  dplyr::summarise_all(class) %>% 
  tidyr::gather(variable, class)

# Model categorical variables as random to report variance fractions
formula_for_variance_fractions = sapply(1:length(col_type$class), function(i){
  switch(col_type$class[i], 
    "factor" = paste0('(1|', col_type$variable[i], ')'),
    "numeric" = col_type$variable[i])
  })

formula_for_variance_fractions <- glue::glue("~ ", glue::glue_collapse(formula_for_variance_fractions, sep = " + "))

violin_plot <- variancePartition::fitExtractVarPartModel(VOOM.GENE_EXPRESSION,
                                                         formula = formula_for_variance_fractions,
                                                         data = COVARIATES)
plotVarPart(sortCols(violin_plot))

# Residuals after normalisation
Dx.Tissue.Gender_RESIDUAL.GENE_EXPRESSION = residuals(Dx.Tissue.Gender_ADJUSTED.FIT, CQN.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(Dx.Tissue.Gender_RESIDUAL.GENE_EXPRESSION, COVARIATES,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```

### Sanity check
```{r residualSigCovars.manual-gender, fig.width=12, fig.height=8}
residualSigCovars[["PC_res"]][[2]]$plotData
```

### Differential expression analysis (Gender-specific SCZ - Control - Other)
Genes that are differentially expressed at an FDR <= 0.05 are:
```{r set-formula-gender}
formula <- glue::glue("~ 0 + ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")
```

`r formula`

```{r diffExp-gender, fig.height=10, fig.width=15}
L1 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.Tissue.GenderSCZ_ACC_Female", "Dx.Tissue.GenderControl_ACC_Female"))
L2 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.Tissue.GenderSCZ_ACC_Male", "Dx.Tissue.GenderControl_ACC_Male"))
L3 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.Tissue.GenderSCZ_DLPFC_Female", "Dx.Tissue.GenderControl_DLPFC_Female"))
L4 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Dx.Tissue.GenderSCZ_DLPFC_Male", "Dx.Tissue.GenderControl_DLPFC_Male"))

L = cbind(L1, L2, L3, L4)

contrast <- c(glue::glue_collapse(names(L1[L1 != 0]), "-"), 
              glue::glue_collapse(names(L2[L2 != 0]), "-"),
              glue::glue_collapse(names(L3[L3 != 0]), "-"),
              glue::glue_collapse(names(L4[L4 != 0]), "-")
)

# Visualize contrast matrix
plotContrasts(L) 

# Fit contrast
FIT.CONTR = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION,
                                     formula = formula,
                                     data = COVARIATES,
                                     L = L)

# Get differential expression
DE_2 = lapply(glue::glue("L{1:4}"), function(i, FIT){
  topTable(FIT, coef=i, number = Inf) %>%
    rownameToFirstColumn('gene_id') %>% 
    left_join(backgroundGenes)
}, FIT.CONTR) 
names(DE_2) <- contrast

DE_2 = DE_2%>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Dx.Tissue.Gender','',Comparison),
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>% 
  tidyr::separate(Comparison, into = c('to.state','ref.state'), sep = '-') %>%
  tidyr::separate(to.state, into = c('to.Dx', 'to.Tissue', 'to.Gender'), sep = '_') %>%
  tidyr::separate(ref.state, into = c('ref.Dx', 'ref.Tissue', 'ref.Gender'), sep = '_') %>% 
  left_join(Ensemble2HGNC) %>%
  left_join(gene_lengths %>% rownameToFirstColumn('gene_id'))
DE_2$Direction[DE_2$adj.P.Val > 0.05 | abs(DE_2$logFC) < log2(1.2)] = 'NONE'

tmp = DE_2 %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, to.Dx, to.Tissue, to.Gender, ref.Dx, ref.Tissue, ref.Gender, logFC) %>%
  group_by(to.Dx, to.Tissue, to.Gender, ref.Dx, ref.Tissue, ref.Gender) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id)))

tmp1 = DE_2 %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, to.Dx, to.Tissue, to.Gender, ref.Dx, ref.Tissue, ref.Gender, logFC) %>%
  group_by(to.Dx, to.Tissue, to.Gender, ref.Dx, ref.Tissue, ref.Gender) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id)))

kable(full_join(tmp,tmp1))

p = ggplot(DE_2, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme_bw() %+replace% theme(legend.position = 'top')
p = p + facet_grid(ref.Tissue~.+ref.Gender, scales = 'fixed')
p = p + ggtitle("Differential Expression between Schizophrenia and Control")
p
```

### Normalisation - Tissue
1. Tissue is chosen as the primary variable of interest (i.e., covariate adjustments is conditioned on diagnosis and reported gender)
2. Individual_ID is chosen as random effect
```{r iterative.normalisation-tissue, results='asis'}
###we might not need this adjustment
COVARIATES <- tidyr::separate(COVARIATES, Dx.Tissue.Gender, c("Dx", "Tissue", "Reported_Gender")) %>% 
  dplyr::mutate(Dx = factor(Dx, levels = c("SCZ", "Control", "Other")),
                Tissue = factor(Tissue), 
                Reported_Gender = factor(Reported_Gender))

adjust.covars[1] <- "Dx"

adjust.covars <- c(adjust.covars, "Tissue", "Reported_Gender")

writeLines(paste('Using following covariates in the model:',
                 paste(adjust.covars, collapse=', '),
                 'as fixed effects and Individual_ID is chosen as random effect'))

# Post adjusted formula
random_effect <- "Individual_ID"
formula <- glue::glue("~ ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")

# Estimate voom weights with DREAM
geneExpr = DGEList(NEW.COUNTS)
geneExpr = edgeR::calcNormFactors(geneExpr)

VOOM.GENE_EXPRESSION = variancePartition::voomWithDreamWeights(counts = geneExpr, 
                                                               formula = formula,
                                                               data = COVARIATES,
                                                               save.plot = TRUE)

plot(VOOM.GENE_EXPRESSION)

# Fit linear model using new weights and new design
VOOM.GENE_EXPRESSION$E = CQN.GENE_EXPRESSION$E
Tissue_ADJUSTED.FIT = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION, 
                                        formula = formula,
                                        data = COVARIATES,
                                        computeResiduals = TRUE)

col_type <- dplyr::select(COVARIATES, c(adjust.covars, "Individual_ID")) %>%
  dplyr::summarise_all(class) %>% 
  tidyr::gather(variable, class)

# Model categorical variables as random to report variance fractions
formula_for_variance_fractions = sapply(1:length(col_type$class), function(i){
  switch(col_type$class[i], 
    "factor" = paste0('(1|', col_type$variable[i], ')'),
    "numeric" = col_type$variable[i])
  })

formula_for_variance_fractions <- glue::glue("~ ", glue::glue_collapse(formula_for_variance_fractions, sep = " + "))

violin_plot <- variancePartition::fitExtractVarPartModel(VOOM.GENE_EXPRESSION,
                                                         formula = formula_for_variance_fractions,
                                                         data = COVARIATES)
plotVarPart(sortCols(violin_plot))

# Residuals after normalisation
Tissue_RESIDUAL.GENE_EXPRESSION = residuals(Tissue_ADJUSTED.FIT, CQN.GENE_EXPRESSION$E)

# Find PC of residual gene expression and significant covariates that are highly correlated with PCs
residualSigCovars = runPCAandPlotCorrelations(Tissue_RESIDUAL.GENE_EXPRESSION, COVARIATES,
                                              'adjusted design(voom-normalized)',
                                              isKeyPlot=TRUE)
```

### Sanity check
```{r residualSigCovars.manual-tissue, fig.width=12, fig.height=8}
residualSigCovars[["PC_res"]][[2]]$plotData
```

### Differential expression analysis (ACC - DLPFC)
Genes that are differentially expressed at an FDR <= 0.05 are:
```{r set-formula-tissue}
formula <- glue::glue("~ 0 + ", glue::glue_collapse(adjust.covars, sep = " + ")) %>% 
  glue::glue(" + (1|{random_effect})")
```

`r formula`

```{r diffExp-tissue, fig.height=10, fig.width=15}
L1 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("TissueDLPFC"))
L2 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("DxSCZ", "DxControl"))
L3 = variancePartition::getContrast(exprObj = VOOM.GENE_EXPRESSION, 
                                   formula = formula, 
                                   data = COVARIATES, 
                                   coefficient = c("Reported_GenderMale"))

L = cbind(L1, L2, L3)

contrast <- c(glue::glue_collapse(names(L1[L1 != 0]), "-"), 
              glue::glue_collapse(names(L2[L2 != 0]), "-"),
              glue::glue_collapse(names(L3[L3 != 0]), "-")
)

# Visualize contrast matrix
plotContrasts(L) 

# Fit contrast
FIT.CONTR = variancePartition::dream(exprObj = VOOM.GENE_EXPRESSION,
                                     formula = formula,
                                     data = COVARIATES,
                                     L = L)

# Get differential expression
DE_3 = lapply(glue::glue("L{1:3}"), function(i, FIT){
  topTable(FIT, coef=i, number = Inf) %>%
    rownameToFirstColumn('gene_id') %>% 
    left_join(backgroundGenes)
}, FIT.CONTR) 
names(DE_3) <- contrast

DE_3 = DE_3%>% 
  rbindlist(idcol = 'Comparison') %>%
  dplyr::mutate(Comparison = gsub('Dx.Tissue','',Comparison),
                Direction = logFC/abs(logFC),
                Direction = factor(Direction, c(-1,1), c('-1' = 'DOWN', '1' = 'UP')),
                Direction = as.character(Direction)) %>% 
  dplyr::mutate(Comparison = ifelse(Comparison %in% "Reported_GenderMale", "Female-Male", Comparison),
                Comparison = ifelse(Comparison %in% "DxSCZ-DxControl", "SCZ-Control", Comparison),
                Comparison = ifelse(Comparison %in% "TissueDLPFC", "ACC-DLPFC", Comparison)) %>% 
  left_join(Ensemble2HGNC) %>%
  left_join(gene_lengths %>% rownameToFirstColumn('gene_id'))
DE_3$Direction[DE_3$adj.P.Val > 0.05 | abs(DE_3$logFC) < log2(1.2)] = 'NONE'

tmp = DE_3 %>%
  dplyr::filter(adj.P.Val <= 0.05) %>%
  dplyr::select(ensembl_gene_id, Comparison) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05 = length(unique(ensembl_gene_id)))

tmp1 = DE_3 %>%
  dplyr::filter(adj.P.Val <= 0.05, abs(logFC) >= log2(1.2)) %>%
  dplyr::select(ensembl_gene_id, Comparison) %>%
  group_by(Comparison) %>%
  dplyr::summarise(FDR_0_05_FC_1.2 = length(unique(ensembl_gene_id)))

kable(full_join(tmp,tmp1))

p = ggplot(DE_3, aes(y = -log10(adj.P.Val), x = logFC, color = Direction)) + geom_point() + xlim(c(-1,1))
p = p + scale_color_manual(values = c('green','grey','red')) + theme_bw() %+replace% theme(legend.position = 'top')
p = p + facet_grid(Comparison~., scales = 'fixed')
p
```


### Store files in synapse
```{r synapse.store, include=FALSE, eval=TRUE, cache=FALSE}
# Set annotations
all.annotations = list(
  dataType = 'geneExpression',
  dataSubtype = 'processed',
  assay	 = 'rnaSeq',
  
  tissue	= 'dorsolateral prefrontal cortex, anterior cingulate cortex', 
  study = 'CMC', 

  species = 'Human',
  consortium	= 'CMC',
   
  analysisType	= "data normalization",
  
  transcriptNormalizationMethod	= 'CQN',
  transcriptQuantificationMethod = 'featureCounts',
  referenceSet = 'GRCh38'
)

# Code
CODE <- Folder(name = "MSSM-Penn-Pitt - ACC and DLPFC - DREAM - FEATURECOUNTS - GRCh38 - CQN", parentId = parentId)
CODE <- synStore(CODE)

# Store covariates
COVARIATES = rownameToFirstColumn(COVARIATES, 'SampleID')
write.table(COVARIATES, file = 'CMC_ACC_DLPFC_Covariates.tsv', sep = '\t', row.names=F, quote=F)
COV_OBJ = synapser::File('CMC_ACC_DLPFC_Covariates.tsv', name = 'Covariates', parentId = CODE$properties$id)
COV_OBJ = synStore(COV_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                      executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(COV_OBJ, annotations = all.annotations)

# Store filtered counts
PROCESSED_COUNTS$filteredExprMatrix$counts %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_Counts.tsv', sep = '\t', row.names=F, quote=F)
COUNT_OBJ = File('CMC_ACC_DLPFC_Counts.tsv', name = 'Counts (filtered raw)', parentId = CODE$properties$id)
COUNT_OBJ = synStore(COUNT_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                      executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(COUNT_OBJ, annotations = all.annotations)

# Store logCPM
CQN.GENE_EXPRESSION$E %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_logCPM.tsv', sep = '\t', row.names=F, quote=F)
LCOUNT_OBJ = File('CMC_ACC_DLPFC_logCPM.tsv', name = 'Counts (filtered logCPM)', parentId = CODE$properties$id)
LCOUNT_OBJ = synStore(LCOUNT_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                      executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(LCOUNT_OBJ, annotations = all.annotations)

# Store cqn offsets
CQN.GENE_EXPRESSION$offset %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_offset.tsv', sep = '\t', row.names=F, quote=F)
OFFSET_OBJ = File('CMC_ACC_DLPFC_offset.tsv', name = 'Gene length and GC content offset', parentId = CODE$properties$id)
OFFSET_OBJ = synStore(OFFSET_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                      executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(OFFSET_OBJ, annotations = all.annotations)

# Store residual gene expression for network analysis - Dx.Tissue
Dx.Tissue_RESIDUAL.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_Dx.Tissue_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
nEXP_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue_netResidualExpression.tsv', 
                name = 'Dx.Tissue Normalised, covariates removed residual expression (for network analysis)', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(nEXP_OBJ, annotations = all.annotations)

# Store residual gene expression for network analysis - Dx.Tissue.Gender
Dx.Tissue.Gender_RESIDUAL.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_Dx.Tissue.Gender_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
nEXP_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue.Gender_netResidualExpression.tsv', 
                name = 'Dx.Tissue.Gender Normalised, covariates removed residual expression (for network analysis)', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(nEXP_OBJ, annotations = all.annotations)

# Store residual gene expression for network analysis - Tissue
Tissue_RESIDUAL.GENE_EXPRESSION %>%
  rownameToFirstColumn('ensembl_gene_id') %>%
  write.table(file = 'CMC_ACC_DLPFC_Tissue_netResidualExpression.tsv', sep = '\t', row.names=F, quote=F)
nEXP_OBJ = File('CMC_ACC_DLPFC_Tissue_netResidualExpression.tsv', 
                name = 'Tissue Normalised, covariates removed residual expression (for network analysis)', 
                parentId = CODE$properties$id)
nEXP_OBJ = synStore(nEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(nEXP_OBJ, annotations = all.annotations)

# Store differential expression results - Dx.Tissue
write.table(DE, file = 'CMC_ACC_DLPFC_Dx.Tissue_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue_DiffExpression.tsv', 
                name = 'Differential Expression Results (Dx.Tissue)', 
                parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(DEXP_OBJ, annotations = all.annotations)

# Store differential expression results - Dx.Tissue.Gender
write.table(DE_2, file = 'CMC_ACC_DLPFC_Dx.Tissue.Gender_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue.Gender_DiffExpression.tsv', 
                name = 'Differential Expression Results (Dx.Tissue.Gender)', 
                parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(DEXP_OBJ, annotations = all.annotations)

# Store differential expression results - Tissue
write.table(DE_3, file = 'CMC_ACC_DLPFC_Tissue_DiffExpression.tsv', sep = '\t', row.names=F, quote=F)
DEXP_OBJ = File('CMC_ACC_DLPFC_Tissue_DiffExpression.tsv', 
                name = 'Differential Expression Results (Tissue)', 
                parentId = CODE$properties$id)
DEXP_OBJ = synStore(DEXP_OBJ, used = ALL_USED_IDs, activityName = activityName, 
                    executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(DEXP_OBJ, annotations = all.annotations)

# Store adjusted fit linear mixed model object as RDS - Dx.Tissue
saveRDS(Dx.Tissue_ADJUSTED.FIT, file = 'CMC_ACC_DLPFC_Dx.Tissue_dream-output.rds')
ADJ_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue_dream-output.rds',
               name = 'Dream Output object - Dx.Tissue', 
               parentId = CODE$properties$id)
ADJ_OBJ = synStore(ADJ_OBJ, used = ALL_USED_IDs, activityName = activityName,
                   executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(ADJ_OBJ, annotations = all.annotations)

# Store adjusted fit linear mixed model object as RDS - Dx.Tissue.Gender
saveRDS(Dx.Tissue.Gender_ADJUSTED.FIT, file = 'CMC_ACC_DLPFC_Dx.Tissue.Gender_dream-output.rds')
ADJ_OBJ = File('CMC_ACC_DLPFC_Dx.Tissue.Gender_dream-output.rds',
               name = 'Dream Output object - Dx.Tissue.Gender', 
               parentId = CODE$properties$id)

ADJ_OBJ = synStore(ADJ_OBJ, used = ALL_USED_IDs, activityName = activityName,
                   executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(ADJ_OBJ, annotations = all.annotations)

# Store adjusted fit linear mixed model object as RDS - Tissue
saveRDS(Dx.Tissue.Gender_ADJUSTED.FIT, file = 'CMC_ACC_DLPFC_Tissue_dream-output.rds')
ADJ_OBJ = File('CMC_ACC_DLPFC_Tissue_dream-output.rds',
               name = 'Dream Output object - Tissue', 
               parentId = CODE$properties$id)
ADJ_OBJ = synStore(ADJ_OBJ, used = ALL_USED_IDs, activityName = activityName,
                   executed = thisFile, activityDescription = activityDescription)
synSetAnnotations(ADJ_OBJ, annotations = all.annotations)
```

|  *Results*                                  |  *SynapseID*                     |
|  -----------------------------------------  |   ---------                      |
|  Covariates                                 |  `r COV_OBJ$properties$id`       |
|  Counts (raw)                               |  `r COUNT_OBJ$properties$id`     |
|  Counts (lcpm)                              |  `r LCOUNT_OBJ$properties$id`    |
|  Offset (for gene length and gc content)    |  `r OFFSET_OBJ$properties$id`    |
|  Design Matrix                              |  `r DM_OBJ$properties$id`        |
|  Residual Expression (for network analysis) |  `r nEXP_OBJ$properties$id`      |
|  Differential Expression                    |  `r DEXP_OBJ$properties$id`      |
|  Enrichment Analysis                        |  `r ENRICH_OBJ$properties$id`    |

### Source Rmd
[Source markdown](`r thisFile`)
